.github/workflows/lint-all.yaml

```diff
@@ -0,0 +1,242 @@
+          filters: |
+            go:
+              - '**/*.go'
+              - '**/go.mod'
+              - '**/go.sum'
+            godeps:
+              - '**/go.mod'
+              - '**/go.sum'
+            helm:
+              - examples/chart/**
+            rust:
+              - '**/Cargo.lock'
+              - '**/Cargo.toml'
+              - '**/*.rs'
+            proto:
+              - '**/*.proto'
+            web:
+              - 'web/**'
+            terraform:
+              - '**.tf'
+              - '**.tf.json'
+              - '**.hcl'
+            sh:
+              - '**/*.sh'
+              - '**/*.sh.tmpl'
+
+  lint-terraform:
+    needs: changes
+    if: ${{ needs.changes.outputs.terraform == 'true' }}
+    uses: gravitational/shared-workflows/.github/workflows/terraform-lint.yaml@main
+    permissions:
+      actions: read
+      contents: read
+      pull-requests: write
+      security-events: write
+
+  lint-rust:
+    name: Lint (Rust)
+    needs: changes
+    if: ${{ needs.changes.outputs.rust == 'true' }}
+    runs-on: ubuntu-22.04-16core
+
+    permissions:
+      contents: read
+
+    container:
+      image: ghcr.io/gravitational/teleport-buildbox:teleport14
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        # TODO(codingllama): Consider https://github.com/actions-rs/clippy-check
+      - name: Run Rust linter
+        run: make lint-rust
+
+  lint-license:
+    name: Check license
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    container:
+      image: ghcr.io/gravitational/teleport-buildbox:teleport14
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Run license linter
+        run: make lint-license
+
+  lint-helm:
+    name: Lint (helm)
+    needs: changes
+    if: ${{ needs.changes.outputs.helm == 'true' }}
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    container:
+      image: ghcr.io/gravitational/teleport-buildbox:teleport14
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Run helm linter
+        run: make lint-helm
+
+  lint-web:
+    name: Lint (UI)
+    needs: changes
+    if: ${{ needs.changes.outputs.web == 'true' }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout OSS Teleport
+        uses: actions/checkout@v4
+      - name: Determine Toolchain Versions
+        run: |
+          echo NODE_VERSION=$(make -s -C build.assets print-node-version) >> $GITHUB_ENV
+      - name: Print versions
+        run: |
+          echo "node: ${NODE_VERSION}"
+      - name: Setup Node Toolchain
+        uses: actions/setup-node@v3
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+      - name: Install Yarn dependencies
+        run: yarn --frozen-lockfile
+      - name: Run lint
+        run: yarn lint
+      - name: Run Type Check
+        run: yarn type-check
```

Comment:
Other than adding `yarn storybook-smoke-test` introduced in #35045, could you also make type-check run before lint?

> I reordered the jobs so that the type check runs before Prettier and ESLint. Type checking is arguably more important – the information that the code is formatted correctly and adheres to project conventions is useless if it fundamentally doesn't type check.

https://github.com/gravitational/teleport/pull/35045#discussion_r1407968996